<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>week1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="Week1_files/libs/clipboard/clipboard.min.js"></script>
<script src="Week1_files/libs/quarto-html/quarto.js"></script>
<script src="Week1_files/libs/quarto-html/popper.min.js"></script>
<script src="Week1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Week1_files/libs/quarto-html/anchor.min.js"></script>
<link href="Week1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Week1_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Week1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Week1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Week1_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<section id="week-1.-single-cell-technology-from-samples-to-raw-data-to-count-matrix" class="level3">
<h3 class="anchored" data-anchor-id="week-1.-single-cell-technology-from-samples-to-raw-data-to-count-matrix">Week 1. Single cell technology, from samples to raw data to count matrix</h3>
<section id="connect-to-the-server" class="level4">
<h4 class="anchored" data-anchor-id="connect-to-the-server">1. Connect to the server</h4>
<p>Find your assigned server name on <a href="https://biohpc.cornell.edu/ww/machines.aspx?i=165">this website</a>. Instructions to connect to the server can be found on <a href="https://biohpc.cornell.edu/lab/doc/remote_access.pdf">this page</a>.</p>
</section>
<section id="set-up-your-working-directory-and-copy-data-files-for-this-exercise." class="level4">
<h4 class="anchored" data-anchor-id="set-up-your-working-directory-and-copy-data-files-for-this-exercise.">2. Set up your working directory and copy data files for this exercise.</h4>
<p><i>Use the ‘copy’ button in the top right corner of each code block to copy these commands.</i></p><i>
</i><p><i>On Windows, use <code>shift-\</code> (hold down shift, then press the backslash key) to paste into the Linux terminal. On a Mac, use <code>command-v</code>.</i></p>
<pre><code>mkdir /workdir/$USER</code></pre>
<pre><code>cd /workdir/$USER</code></pre>
<pre><code>cp -r /workdir/sc_workshop_2024/cellranger .</code></pre>
<details>
<summary>
Expand for Linux details
</summary>
<section id="what-do-those-commands-mean" class="level5">
<h5 class="anchored" data-anchor-id="what-do-those-commands-mean">What do those commands mean?</h5>
<ul>
<li><code>mkdir</code> make directory</li>
<li><code>$USER</code> replace with your netid (i.e.&nbsp;BioHPC username)</li>
<li><code>cd</code> change directory</li>
<li><code>cp -r</code> copy directory and all of its contents to <code>.</code> (dot = here, your current location)</li>
</ul>
</section>
<section id="a-few-more-examples" class="level5">
<h5 class="anchored" data-anchor-id="a-few-more-examples">A few more examples:</h5>
<ul>
<li>Where am I?</li>
</ul>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>pwd</code></p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;= path of the present working directory</p>
<ul>
<li>What files/directories are here?</li>
</ul>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>ls -l</code></p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;= list (long format)</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;<i>you should now see the ‘cellranger’ directory copied to your working directory</i></p>
</section>
<section id="we-will-use-a-lot-of-paths-describing-the-location-of-files" class="level5">
<h5 class="anchored" data-anchor-id="we-will-use-a-lot-of-paths-describing-the-location-of-files">We will use a lot of ‘paths’ describing the location of files</h5>
<ul>
<li>/workdir/$USER: is the path to the working directory you set up for this workshop on your assigned server</li>
<li>For many commands, you can use the full path (starting from the root /) or a relative path, but either way it needs to be correct</li>
</ul>
</section>
<section id="pay-close-attention-to-syntax-and-characters" class="level5">
<h5 class="anchored" data-anchor-id="pay-close-attention-to-syntax-and-characters">Pay close attention to syntax and characters!</h5>
<ul>
<li>forward-slashes denote nested directory and file names</li>
<li>back-slashes have a very different meaning in Linux, they define ‘escape’ sequences to override literal character meanings. A common use of backslashes in this workshop will be to ‘escape’ the end-of-line character when code breaks across several lines. If the code is on a single line, the backslashes are not needed.</li>
</ul>
</section>
<section id="for-more-help-with-how-to-run-a-program-and-optional-parameters" class="level5">
<h5 class="anchored" data-anchor-id="for-more-help-with-how-to-run-a-program-and-optional-parameters">For more help with how to run a program and optional parameters</h5>
<ul>
<li>try <code>program-name --help</code> on the command line.</li>
<li>go to the BioHPC <a href="https://biohpc.cornell.edu/lab/userguide.aspx?a=software">software page</a></li>
</ul>
</section></details>
</section>
</section>
<section id="run-cellranger-to-process-single-cell-rnaseq-data" class="level4">
<h4 class="anchored" data-anchor-id="run-cellranger-to-process-single-cell-rnaseq-data">3. Run cellranger to process single cell RNAseq data</h4>
<p>As it could take several hours to run cellranger on the original data files, special fastq files were prepared with 1/2000 downsampling. Within each sub-directory, you will find a README file with a description of the sample and a link to the original data.</p>
<p>Under the data directory you just copied, there are 6 sub-directories containing fastq files:</p>
<ol type="1">
<li>IgG1d: <strong>Singleplex</strong> single cell gene expression library from GEO series GSE201999.</li>
<li>IgG4: <strong>Singleplex</strong> single cell gene expression library from GEO series GSE201999.</li>
<li>UT: <strong>Singleplex</strong> single cell gene expression library from GEO series GSE201999.</li>
<li>cellplex: <strong>Multiplex</strong> single cell gene expression library.</li>
<li>cellplex_fb: <strong>Multiplex</strong> single cell gene expression library with <strong>antibody capture</strong></li>
<li>refdata-gex-GRCh38-2020-A: <strong>human reference genome required for cellranger</strong></li>
</ol>
<details>
<summary>
Expand for file details
</summary>
<section id="geo-series-gse201999" class="level5">
<h5 class="anchored" data-anchor-id="geo-series-gse201999">GEO series GSE201999</h5>
<p>The first 3 directories contain data from an experiment described in this <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10580117/">paper</a> and will comprise the main dataset used in this workshop.</p>
</section>
<section id="multiplex-datasets" class="level5">
<h5 class="anchored" data-anchor-id="multiplex-datasets">Multiplex datasets</h5>
<p>Two additional datasets are provided as examples to run ‘cellranger multi’ on datasets with hashed (multiplexed) samples and feature barocoding (CITE-seq, antibody capture).</p>
</section>
<section id="human-reference-genome" class="level5">
<h5 class="anchored" data-anchor-id="human-reference-genome">Human reference genome</h5>
<p>This directory was downloaded from 10x Genomics as a pre-indexed reference genome (+transcriptome) in the format required by cellranger. 10x Genomics provides a limited set of pre-indexed genomes; additional indexed genomes can be created with ‘cellranger mkref’ for gene expression (GEX) analysis of other species.</p>
</section>
<section id="what-files-are-located-in-these-directories" class="level5">
<h5 class="anchored" data-anchor-id="what-files-are-located-in-these-directories">What files are located in these directories?</h5>
<ul>
<li>Try listing the contents of each directory</li>
<li>Add a new parameter to list all of the contents of all subdirectories in one command: &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>ls -lR</code> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;R = recursive list</li>
</ul>
</section></details>
</section>
<section id="run-cellranger-count-on-a-singleplex-data-set." class="level5">
<h5 class="anchored" data-anchor-id="run-cellranger-count-on-a-singleplex-data-set.">3.1 Run “cellranger count” on a singleplex data set.</h5>
<pre><code>cd /workdir/$USER/cellranger</code></pre>
<pre><code>export PATH=/programs/cellranger-7.2.0:$PATH</code></pre>
<pre><code>cellranger count --id=run_IgG1d --sample=IgG1d --transcriptome=refdata-gex-GRCh38-2020-A --fastqs=IgG1d  --localcores=8 --localmem=24</code></pre>
<details>
<summary>
Expand for Linux details
</summary>
</details></section>
<section id="what-do-these-commands-mean" class="level5">
<h5 class="anchored" data-anchor-id="what-do-these-commands-mean">What do these commands mean?</h5>
<ul>
<li><p>cd: change directory</p>
<ul>
<li>You were at the top level of /workdir/$USER/, and moved into the subdirectory named ‘cellranger’</li>
<li>You need to be in the ‘cellranger’ subdirectory for the code to work.</li>
<li>If you are not in the right place, figure out where you are (‘pwd’) and move (‘cd’) to /workdir/$USER/cellranger</li>
</ul></li>
<li><p>export PATH: this makes sure your code can find the cellranger program easily</p></li>
<li><p>cellranger count: starts the ‘cellranger count’ program, with the following parameters:</p>
<ul>
<li><code>--id=</code> &nbsp;&nbsp;&nbsp;&nbsp;name of the cellranger run <b>required</b></li>
<li><code>--sample=</code> &nbsp;&nbsp;&nbsp;&nbsp;name of the sample to analyze <b> required, must match the beginning of the fastq filename before the <em>S#</em>)</b></li>
<li><code>--transcriptome=</code> &nbsp;&nbsp;&nbsp;&nbsp;name of the directory containing the reference genome (index formatted for cellranger) <b>required</b></li>
<li><code>--fastqs=</code> &nbsp;&nbsp;&nbsp;&nbsp;must match the directory name containing the fastq files (which happens to be the same as the sample name) <b>required</b></li>
<li><code>--localcores=</code> &nbsp;&nbsp;&nbsp;&nbsp;number of CPUs (cores) that your run of cellranger can use <b>recommended for shared servers</b></li>
<li><code>--localmem=</code> &nbsp;&nbsp;&nbsp;&nbsp;amount of memory that your run of cellranger can use <b>recommended for shared servers</b></li>
</ul></li>
</ul>

<p>Notes:</p>
<ul>
<li><p>This step takes about 5-10 minutes. While you are waiting, you can do Step 4: Run Loupe browser;</p></li>
<li><p>The output files are in the directory run_IgG1d/out. You can browse the output directory. Files of interest include :</p>
<ul>
<li><p>web_summary.html: a summery web page</p></li>
<li><p>possorted_genome_bam.bam: position-sorted read alignment;</p></li>
<li><p>filtered_feature_bc_matrix.h5: HDF5 formatted single gene expression count matrix;</p></li>
<li><p>filtered_feature_bc_matrix: directory containing MEX formatted matrix files.</p>
<p>Either filtered_feature_bc_matrix.h5 or filtered_feature_bc_matrix can be used for downstream data analysis with Seurat or Scanpy.</p></li>
</ul></li>
<li><p>The parameters “–localcores=8 –localmem=24” restrict the memory cpu core usage by the job.</p></li>
<li><p>Cellranger results from the original data of the same library are provided in the directory “/workdir/sc_workshop_2024/GSE201999_output”. In Step 4, you will examine the full dataset (without downsampling).</p></li>
</ul>
</section>
<section id="run-cellranger-multi-on-a-multiplex-data-set" class="level5">
<h5 class="anchored" data-anchor-id="run-cellranger-multi-on-a-multiplex-data-set">3.2 Run “cellranger multi” on a multiplex data set</h5>
<section id="inspect-and-modify-the-cellplexconfig.ori.csv-file" class="level6">
<h6 class="anchored" data-anchor-id="inspect-and-modify-the-cellplexconfig.ori.csv-file">3.2.1 Inspect and modify the cellplex/config.ori.csv file</h6>
<p>Replace all instances of “xxxxx” in the config.ori.csv file with your BioHPC userID, and write the modified content to a new file config.csv. You can use the LINUX sed command to replace the “xxxxx”.</p>
<pre><code>cd /workdir/$USER/cellranger</code></pre>
<pre><code>sed "s/xxxxx/$USER/" cellplex/config.ori.csv &gt; cellplex/config.csv</code></pre>
<p>Confirm that the file is updated with your BioHPC userID with the <code>cat</code> command, which will display the file contents in your terminal.</p>
<pre><code>cat cellplex/config.csv</code></pre>
<p>You can also use a text editor such as <code>nano</code> if you prefer.</p>
</section>
<section id="run-cellranger-multi" class="level6">
<h6 class="anchored" data-anchor-id="run-cellranger-multi">3.2.2 run “cellranger multi”</h6>
<pre><code>cellranger multi --id=run_plex --csv=cellplex/config.csv --localcores=8 --localmem=40</code></pre>
<details>
<summary>
Expand for Linux details
</summary>
</details></section>
</section>
<section id="what-do-these-commands-mean-1" class="level5">
<h5 class="anchored" data-anchor-id="what-do-these-commands-mean-1">What do these commands mean?</h5>
<ul>
<li>cellranger multi: starts the ‘cellranger multi’ program, with the following parameters:
<ul>
<li><code>--id=</code> &nbsp;&nbsp;&nbsp;&nbsp;name of the cellranger run <b>required</b></li>
<li><code>--csv=</code> &nbsp;&nbsp;&nbsp;&nbsp;name of the config file for the run <b>required</b></li>
<li><code>--localcores=</code> &nbsp;&nbsp;&nbsp;&nbsp;number of CPUs (cores) that your run of cellranger can use <b>recommended for shared servers</b></li>
<li><code>--localmem=</code> &nbsp;&nbsp;&nbsp;&nbsp;amount of memory that your run of cellranger can use <b>recommended for shared servers</b></li>
</ul></li>
</ul>
</section>
<section id="how-is-running-cellranger-multi-different-than-cellranger-count" class="level5">
<h5 class="anchored" data-anchor-id="how-is-running-cellranger-multi-different-than-cellranger-count">How is running ‘cellranger multi’ different than ‘cellranger count’?</h5>
<ul>
<li>What information is in the config file vs on the command line?</li>
<li>Use <code>ls</code> to view the output directories and contents. Try adding parameters (-l, -R) to investigate the output directory structure and files.</li>
</ul>

<p>Notes:</p>
<ul>
<li>This step takes about 10-15 minutes.</li>
<li>The output directory is located in run_plex/outs. You would find a “per_sample_outs” directory, with results from each de-multiplexed sample.</li>
</ul>
</section>
<section id="run-cellranger-multi-on-a-multiplex-data-set-with-antibody-capture" class="level5">
<h5 class="anchored" data-anchor-id="run-cellranger-multi-on-a-multiplex-data-set-with-antibody-capture">3.3 Run “cellranger multi” on a multiplex data set with antibody capture</h5>
<section id="inspect-and-modify-the-cellplex_fbconfig.ori.csv-file" class="level6">
<h6 class="anchored" data-anchor-id="inspect-and-modify-the-cellplex_fbconfig.ori.csv-file">3.3.1 Inspect and modify the cellplex_fb/config.ori.csv file</h6>
<p>Replace all “xxxxx” with your BioHPC userID, and write the modified content to a new file config.csv. You can use the LINUX sed command to do this.</p>
<pre><code>cd /workdir/$USER/cellranger</code></pre>
<pre><code>sed "s/xxxxx/$USER/" cellplex_fb/config.ori.csv &gt; cellplex_fb/config.csv</code></pre>
<pre><code>cat cellplex_fb/config.csv</code></pre>
</section>
<section id="run-cellranger-multi-1" class="level6">
<h6 class="anchored" data-anchor-id="run-cellranger-multi-1">3.3.2 run “cellranger multi”</h6>
<pre><code>cellranger multi --id=run_plex_fb --csv=cellplex_fb/config.csv --localcores=8 --localmem=40</code></pre>
<p>Notes:</p>
<ul>
<li>This step takes about 10-15 minutes.</li>
<li>The output directory is located in run_plex_fb/outs. Similar to the previous run, you would find a “per_sample_outs” directory, with results from each demultiplexed sample.</li>
<li>The protein expression levels are included in the gene expression count matrix, as additional features.</li>
</ul>
</section>
</section>

<section id="review-web_summary-qc-and-run-loupe-browser" class="level4">
<h4 class="anchored" data-anchor-id="review-web_summary-qc-and-run-loupe-browser">4. Review ‘web_summary’ QC and run Loupe browser</h4>
<section id="download-the-following-directory-to-your-laptop-or-computer" class="level5">
<h5 class="anchored" data-anchor-id="download-the-following-directory-to-your-laptop-or-computer">4.1 Download the following directory to your laptop or computer</h5>
<p>User Filezilla (or other sftp client) to download cellranger results from the original data.</p>
<p><strong>Host</strong>: your assigned server (cbsuxxxx.biohpc.cornell.edu)</p>
<p><strong>Port:</strong> 22</p>
<p><strong>Remote site</strong>: /workdir/sc_workshop_2024/GSE201999_output</p>
<p><b>Transfer the full directory with Filezilla to your laptop or computer</b></p>
<section id="there-are-three-sub-directories-each-from-a-different-sample" class="level6">
<h6 class="anchored" data-anchor-id="there-are-three-sub-directories-each-from-a-different-sample">There are three sub-directories, each from a different sample:</h6>
<ul>
<li>TL-IgG1</li>
<li>TL-IgG4</li>
<li>Untreated</li>
</ul>
<p>Within each sample directory, there should be a “web_summary.html” file and a “cloupe.cloupe” file.</p>
</section>
</section>
<section id="open-the-web_summary.html-files-in-a-web-browser" class="level5">
<h5 class="anchored" data-anchor-id="open-the-web_summary.html-files-in-a-web-browser">4.2 Open the web_summary.html files in a web browser;</h5>
<p>Documentation of the web summary file and QC metrics:</p>
<p>https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/cr-outputs-web-summary-count https://www.10xgenomics.com/analysis-guides/quality-assessment-using-the-cell-ranger-web-summary</p>
<p>What do the QC metrics indicate about these samples?</p>
</section>
<section id="start-loupe-and-open-the-file-cloupe.cloupe" class="level5">
<h5 class="anchored" data-anchor-id="start-loupe-and-open-the-file-cloupe.cloupe">4.3. Start Loupe, and open the file cloupe.cloupe;</h5>
<p>Tutorials for Loupe:</p>
<p>https://www.10xgenomics.com/support/software/loupe-browser/latest/tutorials/introduction/lb-sc-interface-and-navigation</p>
<p>https://www.10xgenomics.com/support/software/loupe-browser/latest/tutorials/assay-analysis/lb-single-cell-gene-expression-and-flex</p>
</section>
</section>
<section id="things-to-take-into-consideration-when-working-with-real-data" class="level4">
<h4 class="anchored" data-anchor-id="things-to-take-into-consideration-when-working-with-real-data">5. Things to take into consideration when working with real data</h4>
<ul>
<li><p>It could take a few hours to process each set of real data files. You will need to run cellranger in a persistent “screen” session;</p></li>
<li><p>Most likely you would need to run cellranger on multiple samples. Some of the BioHPC servers (large memory gen2) have &gt;100 cpu cores, and there is no performance benefit to run cellranger on a single sample with &gt;32 CPU cores. You would want to run the jobs in parallel to use the all the available CPU cores.</p></li>
</ul>
<p><strong>You can do this step later during the week.</strong></p>
<section id="create-a-batch-script-file-using-nano" class="level5">
<h5 class="anchored" data-anchor-id="create-a-batch-script-file-using-nano">5.1 Create a batch script file using <code>nano</code></h5>
<blockquote class="blockquote">
<p>[!Note] <code>nano</code> is a text editor that is available within terminal. <br>We can use nano to create/edit new or existing files. <a href="https://linuxize.com/post/how-to-use-nano-text-editor/#opening-and-creating-files">Click here for a brief tutorial on using nano</a></p>
</blockquote>
<p>In your terminal write nano followed by the name of the file that you want to write in. In this case we will type:</p>
<pre><code>nano run.sh</code></pre>
<p>This will open a nano editor within your terminal. <br> Copy and paste the following code in your nano editor. On Windows, use <code>shift-\</code> (hold down shift, then press the backslash key) to paste into the Linux terminal. On a Mac, use <code>command-v</code>.</p>
<pre><code>cellranger count --id=run_IgG1d --sample=IgG1d --transcriptome=/workdir/$USER/cellranger/refdata-gex-GRCh38-2020-A --fastqs=/workdir/$USER/cellranger/IgG1d  --localcores=8 --localmem=24

cellranger count --id=run_IgG4 --sample=IgG4 --transcriptome=/workdir/$USER/cellranger/refdata-gex-GRCh38-2020-A --fastqs=/workdir/$USER/cellranger/IgG4  --localcores=8 --localmem=24

cellranger count --id=run_UT --sample=UT --transcriptome=/workdir/$USER/cellranger/refdata-gex-GRCh38-2020-A --fastqs=/workdir/$USER/cellranger/UT  --localcores=8 --localmem=24</code></pre>
<p>Once all three lines are added in your nano editor, hold down <code>Ctrl+X</code> You will see towards the bottom a prompt will appear saying <code>Save modified buffer</code>. Type <code>Y</code> and the prompt will change to <code>File Name to write : run.sh</code>. Hit return/enter on the keybpard and your script is now created. If you type <code>ls</code> in your terminal, you will see there will be a new file called <code>run.sh</code> that is now present in your working directory.</p>
<section id="notes" class="level6">
<h6 class="anchored" data-anchor-id="notes">Notes:</h6>
<ul>
<li>This code uses the full path for the reference index (<code>--transcriptome</code>) and fastq location (<code>--fastqs=</code>), allowing the script to be run from any directory.</li>
<li>The output directories (named as <code>--id=</code>) will be created in the directory that you run the script.</li>
</ul>
</section>
</section>
<section id="run-the-script-in-screen-persistent-session." class="level5">
<h5 class="anchored" data-anchor-id="run-the-script-in-screen-persistent-session.">5.2 Run the script in “screen” persistent session.</h5>
<p>Tutorial for “screen” can be found <a href="https://biohpc.cornell.edu/lab/doc/Linux_exercise_part2.pdf">here</a>.</p>
<p>There are two ways to run this script:</p>
<p>Run in serial (run cellranger on one sample at a time):</p>
<pre><code>export PATH=/programs/cellranger-7.2.0:$PATH
sh run.sh</code></pre>
<p>Run in parallel (“-j 2” : process two samples at a time):</p>
<pre><code>export PATH=/programs/cellranger-7.2.0:$PATH
parallel -j 2 &lt; run.sh</code></pre>
<p>Notes:</p>
<ul>
<li>When running in parallel, make sure that the numbers set in “–localmem –localcores” multiply by number of jobs do not exceed the total RAM or CPU cores on the server you are using. Your assigned server for this workshop has 128GB RAM and 24 CPU cores.</li>
</ul>
</section>
</section>


</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>